morf-core/src/main/java/org/alfasoftware/morf/sql/SqlTokenizer.java

```diff
@@ -0,0 +1,707 @@
+      this.position = from;
+      this.tokenString = query.substring(from, to);
+    }
+
+    public TokenType getType() {
+      return type;
+    }
+
+    public String getString() {
+      return tokenString;
+    }
+
+    public int getPosition() {
+      return position;
+    }
+  }
+
+
+  /**
+   * Type of a token.
+   */
+  public enum TokenType {
+    /**
+     * Sequence of white-spaces, including newlines.
+     */
+    WHITESPACE,
+
+    /**
+     * Single identifier or a keyword.
+     * Note that the tokenizer does not attempt semantic analysis,
+     * therefore keywords are not being distinguished from identifiers.
+     */
+    IDENTIFIER,
+
+    /**
+     * Single number.
+     */
+    NUMBER,
+
+    /**
+     * Double-quoted identifier, including the double-quotes, and any leading classifiers.
+     */
+    DQUOTED,
+
+    /**
+     * Single-quoted string literal, including the single-quotes.
+     */
+    STRING,
+
+    /**
+     * Dollar-quoted string literal, including the dollar-quotes.
+     */
+    DSTRING,
+
+    /**
+     * Single-character operator.
+     */
+    OPERATOR,
+
+    /**
+     * Single-line comment, including the comment-starting characters and comment-ending newline.
+     */
+    SCOMMENT,
+
+    /**
+     * Multi-line comment, including the comment-starting characters and comment-ending characters.
+     */
+    MCOMMENT,
+
+    /**
+     * An illegal or unexpected character character, exactly as found.
+     */
+    ILLEGAL
+  }
+
+
+  /**
+   * Pure static class.
+   */
+  private SqlTokenizer() {
+  }
+
+
+  /**
+   * Tokenizes given query, i.e. splits the query into individual tokens. See {@link TokenType}.
+   *
+   * <p>
+   * The caller can examine the tokens via the provided callback function, which is called once for each token.
+   * The function can, for example, be used to collect all the tokens into a list, or to validate the tokens.
+   * </p>
+   *
+   * <p>
+   * Note: This tokenizer reports white-spaces and comments as tokens, even though they are not really tokens.
+   * </p>
+   *
+   * <p>
+   * All tokens, simply concatenated back together, will always give the original input query,
+   * i.e. no characters are added/removed/translated, no tokens are ever excluded, or reordered.
+   * </p>
+   *
+   * <p>Simple example:</p>
+   * <pre>SELECT * FROM table</pre>
+   * <p>Is tokenized as:</p>
+   * <pre>IDENTIFIER(SELECT), WHITESPACE( ), OPERATOR(*), WHITESPACE( ), IDENTIFIER(FROM), WHITESPACE( ), IDENTIFIER(table)</pre>
+   *
+   * @param databaseType ORACLE/PGSQL
+   * @param query Query to be tokenized.
+   * @param visitor Function which will receive TokenType+String pairs.
+   *                The string contains the entire token, as described by the respective TokenType.
+   */
+  public static void tokenizeSqlQuery(String databaseType, String query, Consumer<Token> visitor) {
+    final int length = query.length();
+    int i = 0;
+
+    while (i < length) {
+
```

Comment:
Do you think it would be worth extracting some of this logic outside the loop here?

i = processNextToken(databaseType, query, length, i, visitor);
^ inside the loop

 private int processNextToken(String databaseType, String query, int length, int i, Consumer<Token> visitor) { ... }
^ the same logic extracted outside the loop